pkg/common-controller/groupsnapshot_controller.go

```diff
@@ -9,3 +22,236 @@ func (ctrl *csiSnapshotCommonController) storeGroupSnapshotUpdate(groupsnapshot
+func (ctrl *csiSnapshotCommonController) getClaimFromVolumeGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (*v1.PersistentVolumeClaim, error) {
+	labelSelector := groupSnapshot.Spec.Source.Selector
+
+	// Get PVC that has group snapshot label applied.
+	pvcList, err := ctrl.client.CoreV1().PersistentVolumeClaims(groupSnapshot.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector.String()})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list PVCs with label selector %s: %q", labelSelector.String(), err)
+	}
+	if len(pvcList.Items) == 0 {
+		return nil, fmt.Errorf("label selector %s for group snapshot not applied to any PVC", labelSelector.String())
+	}
+	return &pvcList.Items[0], nil
+}
+
+// updateGroupSnapshot runs in worker thread and handles "groupsnapshot added",
+// "groupsnapshot updated" and "periodic sync" events.
+func (ctrl *csiSnapshotCommonController) updateGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) error {
+	// Store the new group snapshot version in the cache and do not process it
+	// if this is an old version.
+	klog.V(5).Infof("updateGroupSnapshot %q", utils.GroupSnapshotKey(groupSnapshot))
+	newGroupSnapshot, err := ctrl.storeGroupSnapshotUpdate(groupSnapshot)
+	if err != nil {
+		klog.Errorf("%v", err)
+	}
+	if !newGroupSnapshot {
+		return nil
+	}
+
+	//err = ctrl.syncSnapshot(groupSnapshot)
+	if err != nil {
+		if errors.IsConflict(err) {
+			// Version conflict error happens quite often and the controller
+			// recovers from it easily.
+			klog.V(3).Infof("could not sync snapshot %q: %+v", utils.GroupSnapshotKey(groupSnapshot), err)
+		} else {
+			klog.Errorf("could not sync snapshot %q: %+v", utils.GroupSnapshotKey(groupSnapshot), err)
+		}
+		return err
+	}
+	return nil
+}
+
+// deleteGroupSnapshot runs in worker thread and handles "groupsnapshot deleted" event.
+func (ctrl *csiSnapshotCommonController) deleteGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) {
+	_ = ctrl.snapshotStore.Delete(groupSnapshot)
+	klog.V(4).Infof("snapshot %q deleted", utils.GroupSnapshotKey(groupSnapshot))
+
+	groupSnapshotContentName := ""
+	if groupSnapshot.Status != nil && groupSnapshot.Status.BoundVolumeGroupSnapshotContentName != nil {
+		groupSnapshotContentName = *groupSnapshot.Status.BoundVolumeGroupSnapshotContentName
+	}
+	if groupSnapshotContentName == "" {
+		klog.V(5).Infof("deleteGroupSnapshot[%q]: content not bound", utils.GroupSnapshotKey(groupSnapshot))
+		return
+	}
+
+	// sync the content when its group snapshot is deleted.  Explicitly sync'ing
+	// the content here in response to group snapshot deletion prevents the content
+	// from waiting until the next sync period for its release.
+	klog.V(5).Infof("deleteGroupSnapshot[%q]: scheduling sync of content %s", utils.GroupSnapshotKey(groupSnapshot), groupSnapshotContentName)
+	ctrl.groupContentQueue.Add(groupSnapshotContentName)
+}
```

Comment:
This needs to trigger deletion of individual snapshots.  Add a TODO if it is not there yet.

Reply:
It looks like from here - https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller_base.go#L461 - that this function just adds the object to the content queue and then the reconciler handles deletion when the queue is popped..

Reply:
Also, I tried to keep this PR to deal with only `GroupSnapshot` related functionality and not `GroupSnapshotContent`, but ended up introducing some variables and function templates related to the latter. Should I remove that? It will make the code review process easier. I can introduce all `GroupSnapshotContent` related work in a separate PR

Reply:
It's fine either way.

> It looks like from here - https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller_base.go#L461 - that this function just adds the object to the content queue and then the reconciler handles deletion when the queue is popped.. Should the TODO be here or in that function?

That's fine.